### ISSUE TITLE
Questions about the paper

### ISSUE THREAD

**Issue by guozhiyao** [Jung-Jun-Uk/UNPG#1]
I feel confused about the Fig.2 in the paper. 
1. The `x_1` and `x_3` should be the class 2, while the `x_2` and `x_4` should be the class 4 according to the `CLPG` of the picture. But the right part of picture shows that only the `x_1` and `x_2` are the same class.
2.  Why we should collect the same whole negative pairs for each input? For the input `x_1`, only the `s_2^n` and `s_3^n` are the negative pairs, why we should include the other negative pairs for the `x_1`? When we have huge batch size, it costs a lot of gpu memory.

**Comment by junuke** [comment]
Hi,

Q1. thank you for questioning. it is a mistake, we'll fix it in a later version.
Q2.  The reason is to satisfy WDFS. For explanation, I attach the content of the work currently being edited. our UNPG is designed so that one positive pair of similarity is higher than the whole negative pairs of similarity in minibatch as in eq 10. GPU memory and cost increase in mini-batch 512 (our experimental setup), is negligible
![unpg_question](https://user-images.githubusercontent.com/58498680/160973057-c15032fc-8e22-473c-9778-6a86baf39e00.PNG)
.

**Comment by guozhiyao** [comment]
Thank you for your reply.
About the Q2. We can observe the gradient of Eq.8. For items that contain `x_i` in `MLPG`, `x_i` will be optimized away from `x_j`; while for items that do not contain `x_i`, they will not contribute gradients to the update of `x_i`, but only affect its `logit`. And if the batch size is 512, it will create a 512x130560 matrix of MLPG (The 130560=512*512/2-512, if we do not use the `Noise Negative Pair Filtering` and the samples have different classes in the batch), and it is O(N^3). Have you try that only keep the negative pairs related to `x_i`？

**Comment by junuke** [comment]
I understand the intent of the question. 

1. As you know, the gradient is derived as follows: 
![unpg_question2](https://user-images.githubusercontent.com/58498680/161009671-f253aa7e-c196-490d-bdeb-245710761a8c.PNG)
2. Tricks to manipulate probabilities are important. For example, many margin injecting methods (e.g., CosFace cos\theta+m) also only affect its logit but improve the performance.
3. In my memory, the manner you said is lower performance than UNPG on the face recognition task but improves the image retrieval task (though more tests are needed)
4. Contrary to your concern, UNPG's computation cost is small. There was only a slight difference in memory usage and training speed in our experiments.

**Comment by guozhiyao** [comment]
Thank you for your reply.
Then I can understand your method as introducing more difficult negative pairs to push the similarity of positive pairs to become larger. Have you compared to the margin-based methods with larger margin?
For the usage of gpu memory, maybe my batch size is too large. I try to reduce the batch size first.

**Comment by junuke** [comment]
We appreciate your interest in UNPG.
Margins need to be selected appropriately, and large margins do not guarantee better performance. (See ArcFace) We only experimented with a margin of m=0.5, used in many papers.

**Comment by guozhiyao** [comment]
Thank you for your reply.

**Comment by guozhiyao** [comment]
Sorry to say it again.
I tried a smaller batch size (such as 512), and the GPU usage does not change much, but when the batch size is relatively large (such as 1024), the GPU usage will increase a lot. I use the V100 32G GPU card for training here. The normal cosface takes up about 20G, and the ULPG takes up more than 32G when the batch is up to 1024.
I tried to modify the implementation, which can reduce the GPU usage to `O(N^2)`, you can see if this works.
1. Perform matrix multiplication on the input to obtain the similarity between samples, and generate a mask, where the selected sample pairs are set to 1, and the others are set to 0. It is expressed as `simi`, and the GPU usage is `B*B`, where `B` is the batch size.
2. CLPG is calculated normally, and the similarity matrix `cosine` is obtained, and the size is `B*N`, where `N` is the class number.
3. When calculating the softmax forward propagation, the denominator is `torch.sum(cosine,dim=1)+torch.sum(simi)`, which is consistent with your Eq.8.
4. When calculating backpropagation of softmax, you only need to use `simi`, because for `xi`, the similarity of other negative samples has nothing to do with it, so you don't need to consider it in the backpropagation, you only need to use `simi` to get the gradient of `xi`.
When I implement ULPG in this way, the memory usage is 22G (batch is 1024).

**Comment by junuke** [comment]
Sorry for the delayed answer. 

I had a COVID-19, took a break, and returned to work now. 

It is difficult to understand with text alone, but anyway, it is good if you implement the custom forward and backward steps to avoid allocating a lot of GPU memory.

**Comment by guozhiyao** [comment]
Hope you've recovered.
I feel that there is a certain problem with the above Eq.12. Assuming that there is a sample pair `xi,xj`, which will exist in the denominator of all losses, when get the derivation of `xi`, you cannot only consider the loss of `xi`, but also the loss of other samples.

**Comment by junuke** [comment]
I need some time to answer. I plan to check the gradients by implementing custom backward with unpg, and then I will revise the equation if needed. But not now; I don't have time because I am working on other projects currently. I'll share it later when I make it.
